### ISSUE TITLE
Update memcache client with new commands of memcached 1.2.4

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2957]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2957|
|Type|enhancement|
|Created|2007-12-17 21:50:55Z|
|Branch|https://github.com/twisted/twisted/tree/memcache-update-2957|

There are 4 new commands: gets, append, prepend, cas.

<details><summary>Searchable metadata</summary>

```
trac-id__2957 2957
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__branches_memcache_update_2957 branches/memcache-update-2957
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1197928255000000 1197928255000000
changetime__1201965725000000 1201965725000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [22100]) Branching to 'memcache-update-2957'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22102]) Add the 4 new methods and tests for them.

Refs [#2957](https://github.com/twisted/twisted/issues/2957)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

This is ready to review. The gets and cas facility is not that awesome, but hey.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

For the ```cas``` docstring, I'd say something like this:

```
Change the content of C{key} only if the C{cas} value matches the
current one associated with the key. Use this to store a value which
hasn't been modified since last time you fetched it.
```

Also, capitalize the first letter of the first word of a sentence (eg, for the @param docs). ;)

**cas** is a really opaque name.  Even after reading the diff, I don't know what those letters in that order mean.  Is there something more understandable we can use here?  Likewise, the name ```gets``` doesn't really tell me that it differs from ```get``` in that it returns the _cas_ value as well.

  * _64 bits value_ should probably be _64 bit value_
  * _if the operations has_ should probably be _if the operation has_ (multiple places)
  * setUp docstring misspells _protocol_ as _protoco_

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [22180]) Fix the docstrings.

Refs [#2957](https://github.com/twisted/twisted/issues/2957)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

I'm not sure what to do here. The memcached semantics is not really clear. The cas is a unique numerical identifier, that's all I know. Do you think docstring needs further enhancements?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

I found "cas" on google, it seems to mean "check and set", which is reasonable in context.

So perhaps the ```cas``` method can be renamed ```checkAndSet``` (it should still have _cas_ in the docstring I guess, since most people seem to refer to this feature that way; I'm not sure if they know what it stands for either).

Maybe ```gets``` could be ```getWithIdentifier```, or maybe ```get``` could take a flag that specifies this behavior.  Just random ideas, I don't feel as strongly about this as about the ```cas``` method.  I leave it up to your judgement.

I noticed there seems to be no direct coverage of the failure case for ```append``` or ```prepend```.  This is sort of tested by ```test_invalidCommand```, but direct tests would be good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [22340]) Process review comments.

Refs [#2957](https://github.com/twisted/twisted/issues/2957)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Thanks for the search, I didn't manage to find something but checkAndSet is indeed clear in its goal. I think I've done all points.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

The docstring for `MemCacheProtocol.get` is missing some text in the _@param withIdentifier_ field.

Everything else looks fine, merge when that's fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [22445]) Merge memcache-update-2957

Author: therve
Reviewer: exarkun
Fixes [#2957](https://github.com/twisted/twisted/issues/2957)

Update the memcache client protocol with the new commands of memcached 1.2.4.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
